### ISSUE TITLE
java.lang.NoSuchFieldException for CreateTableAsSelect statements

### ISSUE THREAD

**Issue by afgaron** [AbsaOSS/spline-spark-agent#487]
- Scala v2.12
- Spark v3.2.1
- Spline v0.7.8

Running a command in the style of

```
create table B as (
    select val from A
)
```

throws a `java.lang.NoSuchFieldException`. The `CreateTableAsSelect` command seems like a pretty important one but isn't listed in the spline README anywhere. Full error log below:

```
22/07/29 00:38:29 WARN LineageHarvester: Write extraction failed for: class org.apache.spark.sql.catalyst.plans.logical.CreateTableAsSelect
java.lang.NoSuchFieldException: org.apache.spark.sql.catalyst.plans.logical.CreateTableAsSelect.catalog
	at za.co.absa.commons.reflect.ReflectionUtils$.$anonfun$extractFieldValue$9(ReflectionUtils.scala:175)
	at scala.Option.getOrElse(Option.scala:189)
	at za.co.absa.commons.reflect.ReflectionUtils$.extractFieldValue(ReflectionUtils.scala:175)
	at za.co.absa.commons.reflect.ReflectionUtils$.extractFieldValue(ReflectionUtils.scala:184)
	at za.co.absa.spline.harvester.plugin.embedded.DataSourceV2Plugin.za$co$absa$spline$harvester$plugin$embedded$DataSourceV2Plugin$$processV2CreateTableCommand(DataSourceV2Plugin.scala:104)
	at za.co.absa.spline.harvester.plugin.embedded.DataSourceV2Plugin$$anonfun$2.applyOrElse(DataSourceV2Plugin.scala:73)
	at za.co.absa.spline.harvester.plugin.embedded.DataSourceV2Plugin$$anonfun$2.applyOrElse(DataSourceV2Plugin.scala:52)
	at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:175)
	at scala.PartialFunction$Lifted.apply(PartialFunction.scala:228)
	at scala.PartialFunction$Lifted.apply(PartialFunction.scala:224)
	at za.co.absa.spline.harvester.builder.write.PluggableWriteCommandExtractor.asWriteCommand(PluggableWriteCommandExtractor.scala:44)
	at za.co.absa.spline.harvester.LineageHarvester.$anonfun$tryExtractWriteCommand$1(LineageHarvester.scala:145)
	at scala.util.Try$.apply(Try.scala:213)
	at za.co.absa.spline.harvester.LineageHarvester.tryExtractWriteCommand(LineageHarvester.scala:145)
	at za.co.absa.spline.harvester.LineageHarvester.harvest(LineageHarvester.scala:64)
	at za.co.absa.spline.harvester.QueryExecutionEventHandler.onSuccess(QueryExecutionEventHandler.scala:42)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener.$anonfun$onSuccess$2(SplineQueryExecutionListener.scala:40)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener.$anonfun$onSuccess$2$adapted(SplineQueryExecutionListener.scala:40)
	at scala.Option.foreach(Option.scala:407)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener.$anonfun$onSuccess$1(SplineQueryExecutionListener.scala:40)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener.withErrorHandling(SplineQueryExecutionListener.scala:49)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener.onSuccess(SplineQueryExecutionListener.scala:40)
	at org.apache.spark.sql.util.ExecutionListenerBus.doPostEvent(QueryExecutionListener.scala:165)
	at org.apache.spark.sql.util.ExecutionListenerBus.doPostEvent(QueryExecutionListener.scala:135)
	at org.apache.spark.util.ListenerBus.postToAll(ListenerBus.scala:118)
	at org.apache.spark.util.ListenerBus.postToAll$(ListenerBus.scala:102)
	at org.apache.spark.sql.util.ExecutionListenerBus.postToAll(QueryExecutionListener.scala:135)
	at org.apache.spark.sql.util.ExecutionListenerBus.onOtherEvent(QueryExecutionListener.scala:147)
	at org.apache.spark.scheduler.SparkListenerBus.doPostEvent(SparkListenerBus.scala:102)
	at org.apache.spark.scheduler.SparkListenerBus.doPostEvent$(SparkListenerBus.scala:28)
	at org.apache.spark.scheduler.AsyncEventQueue.doPostEvent(AsyncEventQueue.scala:37)
	at org.apache.spark.scheduler.AsyncEventQueue.doPostEvent(AsyncEventQueue.scala:37)
	at org.apache.spark.util.ListenerBus.postToAll(ListenerBus.scala:118)
	at org.apache.spark.util.ListenerBus.postToAll$(ListenerBus.scala:102)
	at org.apache.spark.scheduler.AsyncEventQueue.super$postToAll(AsyncEventQueue.scala:105)
	at org.apache.spark.scheduler.AsyncEventQueue.$anonfun$dispatch$1(AsyncEventQueue.scala:105)
	at scala.runtime.java8.JFunction0$mcJ$sp.apply(JFunction0$mcJ$sp.java:23)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)
	at org.apache.spark.scheduler.AsyncEventQueue.org$apache$spark$scheduler$AsyncEventQueue$$dispatch(AsyncEventQueue.scala:100)
	at org.apache.spark.scheduler.AsyncEventQueue$$anon$2.$anonfun$run$1(AsyncEventQueue.scala:96)
	at org.apache.spark.util.Utils$.tryOrStopSparkContext(Utils.scala:1623)
	at org.apache.spark.scheduler.AsyncEventQueue$$anon$2.run(AsyncEventQueue.scala:96)
22/07/29 00:38:29 INFO ClusterLoadMonitor: Added query with execution ID:55. Current active queries:1
22/07/29 00:38:29 INFO ClusterLoadMonitor: Removed query with execution ID:55. Current active queries:0
22/07/29 00:38:29 WARN LineageHarvester: 
****************************** OBJECT DUMP BEGIN ******************************
class org.apache.spark.sql.catalyst.plans.logical.CreateTableAsSelect
  resolved: boolean = true
  name: org.apache.spark.sql.catalyst.plans.logical.LogicalPlan 
    catalog: org.apache.spark.sql.connector.catalog.TableCatalog 
      name: java.lang.String = spark_catalog
      sessionCatalog: org.apache.spark.sql.connector.catalog.CatalogPlugin 
        BestEffortStagedTable$module: com.databricks.sql.transaction.tahoe.catalog.DeltaCatalog$BestEffortStagedTable$ = null
        spark: org.apache.spark.sql.SparkSession 
          implicits$module: org.apache.spark.sql.SparkSession$implicits$ 
            $outer: org.apache.spark.sql.SparkSession ! Object was already logged
          org$apache$spark$sql$SparkSession$$creationSite: org.apache.spark.util.CallSite 
            shortForm: java.lang.String = newSession at DriverSparkHooks.scala:191
            longForm: java.lang.String = org.apache.spark.sql.hive.HiveContext.newSession(HiveContext.scala:32)
com.databricks.backend.daemon.driver.DriverSparkHooks.sessionizedDriverHooks(DriverSparkHooks.scala:191)
com.databricks.backend.daemon.driver.DriverCorral.createDriverSparkHooks(DriverCorral.scala:390)
com.databricks.backend.daemon.driver.DriverCorral.$anonfun$addReplToExecutionContext$1(DriverCorral.scala:523)
scala.collection.mutable.HashMap.getOrElseUpdate(HashMap.scala:86)
com.databricks.backend.daemon.driver.DriverCorral.addReplToExecutionContext(DriverCorral.scala:523)
com.databricks.backend.daemon.driver.DriverCorral.com$databricks$backend$daemon$driver$DriverCorral$$handleRequest(DriverCorral.scala:671)
com.databricks.backend.daemon.driver.DriverCorral$$anonfun$receive$1.applyOrElse(DriverCorral.scala:1062)
com.databricks.backend.daemon.driver.DriverCorral$$anonfun$receive$1.applyOrElse(DriverCorral.scala:1058)
com.databricks.rpc.ServerBackend.$anonfun$internalReceive$2(ServerBackend.scala:90)
com.databricks.rpc.ServerBackend$$anonfun$commonReceive$1.applyOrElse(ServerBackend.scala:117)
com.databricks.rpc.ServerBackend$$anonfun$commonReceive$1.applyOrElse(ServerBackend.scala:117)
com.databricks.rpc.ServerBackend.$anonfun$internalReceive$1(ServerBackend.scala:75)
com.databricks.logging.UsageLogging.$anonfun$recordOperation$1(UsageLogging.scala:330)
com.databricks.logging.UsageLogging.executeThunkAndCaptureResultTags$1(UsageLogging.scala:424)
com.databricks.logging.UsageLogging.$anonfun$recordOperationWithResultTags$4(UsageLogging.scala:444)
com.databricks.logging.Log4jUsageLoggingShim$.$anonfun$withAttributionContext$1(Log4jUsageLoggingShim.scala:33)
scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)
com.databricks.logging.AttributionContext$.withValue(AttributionContext.scala:94)
com.databricks.logging.Log4jUsageLoggingShim$.withAttributionContext(Log4jUsageLoggingShim.scala:31)
          currentRole: java.util.concurrent.atomic.AtomicReference 
            value: java.lang.Object = null
          pythonProcessUniquifier: scala.Option = None
          sessionUUID: java.lang.String = 27561b9b-3925-4309-baba-34b148759099
        catalog: org.apache.spark.sql.catalyst.catalog.SessionCatalog 
        bitmap$0: boolean = true
      mcsc: com.databricks.sql.managedcatalog.ManagedCatalogSessionCatalog ! Object was already logged
    ident: org.apache.spark.sql.connector.catalog.Identifier 
      namespace: [Ljava.lang.String; 
      name: java.lang.String = external
  partitioning: scala.collection.Seq = List()
  query: org.apache.spark.sql.catalyst.plans.logical.LogicalPlan 
    resolved: boolean = true
    validConstraints: org.apache.spark.sql.catalyst.expressions.ExpressionSet = ExpressionSet()
    projectList: scala.collection.Seq = List(val1#1775, val2#1776, total#1777, diff#1778)
    nodePatterns: scala.collection.Seq = List(PROJECT)
    bitmap$0: byte = 3
  tableSpec: org.apache.spark.sql.catalyst.plans.logical.TableSpec 
    bucketSpec: scala.Option = None
    properties: scala.collection.immutable.Map = Map()
    provider: scala.Option = None
    options: scala.collection.immutable.Map = Map()
    location: scala.Option Some
      x: java.lang.String = dbfs:/user/hive/warehouse/cma_users_agaron.db/external/
    comment: scala.Option = None
    serde: scala.Option = None
    external: boolean = false
  writeOptions: scala.collection.immutable.Map = Map()
  ignoreIfExists: boolean = false
  output: scala.collection.Seq = List(num_affected_rows#1785L, num_inserted_rows#1786L)
  nodePatterns: scala.collection.Seq = List(COMMAND)
  bitmap$0: boolean = true
****************************** OBJECT DUMP END   ******************************
```

**Comment by cerveada** [comment]
From the log it seems you are using Databricks. Which version of Databricks have you used?

**Comment by afgaron** [comment]
I encountered this issue on runtimes 10.4 and 10.5. It didn't throw an error on 7.3 or 9.1.

**Comment by mohitgarg4628** [comment]
Hi @afgaron 

Are you able to solve this issue. I'm getting same issue on databricks

Thanks

**Comment by juliuszsompolski** [comment]
CreateTableAsSelect doesn't have catalog field anymore since https://github.com/apache/spark/pull/34667/files, which Spline agent is trying to grab through reflection.
Edit: That change is in Spark 3.3 only, but it is also present in Databricks Runtime 10.4 and 10.5, even though they are based on Spark 3.2. Databricks Runtimes internal interfaces do differ from Apache Spark.

**Comment by cerveada** [comment]
> ... Edit: That change is in Spark 3.3 only, but it is also present in Databricks Runtime 10.4 and 10.5, even though they are based on Spark 3.2. Databricks Runtimes internal interfaces do differ from Apache Spark.

Yes, It seems like these changes were made in Databricks first and later merged into Spark 3.3 itself.

I already started working on #508. If my understanding is correct, it should fix this issue as well.

**Comment by juliuszsompolski** [comment]
Thanks.
It looks to me that now CTAS and RTAS have a `name: LogicalPlan` which should be a `ResolvedIdentifier` , and then it has a `catalog: CatalogPlugin`, which can be one of multiple implementations of the catalog... if it's implementing a`TableCatalog` subinterface, you should get what you want... But I'm not really familiar with the DatasourceV2 interfaces, so I don't know more...

**Comment by cerveada** [comment]
fixed by #508
